File: app/src/main/java/com/veganbeauty/app/features/account/order/OrderListAdapter.java
package com.veganbeauty.app.features.account.order;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;


import com.veganbeauty.app.R;
import com.veganbeauty.app.data.local.entities.OrderEntity;
import com.veganbeauty.app.data.local.entities.OrderEntity.OrderItem;
import com.veganbeauty.app.databinding.AccountOrderItemBinding;

import java.util.List;

public class OrderListAdapter extends ListAdapter<OrderEntity, OrderListAdapter.OrderViewHolder> {

    public interface OnOrderClickListener {
        void onClick(OrderEntity order);
    }

    private final OnOrderClickListener onItemClick;
    private final OnOrderClickListener onCancelClick;
    private final OnOrderClickListener onDetailClick;
    private final OnOrderClickListener onReorderClick;
    private final OnOrderClickListener onTrackClick;
    private final OnOrderClickListener onContactClick;
    private final OnOrderClickListener onReviewClick;

    public OrderListAdapter(
            OnOrderClickListener onItemClick,
            OnOrderClickListener onCancelClick,
            OnOrderClickListener onDetailClick,
            OnOrderClickListener onReorderClick,
            OnOrderClickListener onTrackClick,
            OnOrderClickListener onContactClick,
            OnOrderClickListener onReviewClick
    ) {
        super(new OrderDiffCallback());
        this.onItemClick = onItemClick != null ? onItemClick : order -> {};
        this.onCancelClick = onCancelClick != null ? onCancelClick : order -> {};
        this.onDetailClick = onDetailClick != null ? onDetailClick : order -> {};
        this.onReorderClick = onReorderClick != null ? onReorderClick : order -> {};
        this.onTrackClick = onTrackClick != null ? onTrackClick : order -> {};
        this.onContactClick = onContactClick != null ? onContactClick : order -> {};
        this.onReviewClick = onReviewClick != null ? onReviewClick : order -> {};
    }

    @NonNull
    @Override
    public OrderViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        AccountOrderItemBinding binding = AccountOrderItemBinding.inflate(
                LayoutInflater.from(parent.getContext()),
                parent,
                false
        );
        return new OrderViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull OrderViewHolder holder, int position) {
        holder.bind(
                getItem(position),
                onItemClick,
                onCancelClick,
                onDetailClick,
                onReorderClick,
                onTrackClick,
                onContactClick,
                onReviewClick
        );
    }

    public static class OrderViewHolder extends RecyclerView.ViewHolder {

        private final AccountOrderItemBinding binding;
        private boolean isExpanded = false;

        public OrderViewHolder(@NonNull AccountOrderItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(
                OrderEntity order,
                OnOrderClickListener onItemClick,
                OnOrderClickListener onCancelClick,
                OnOrderClickListener onDetailClick,
                OnOrderClickListener onReorderClick,
                OnOrderClickListener onTrackClick,
                OnOrderClickListener onContactClick,
                OnOrderClickListener onReviewClick
        ) {
            Context context = binding.getRoot().getContext();

            // Khôi phục feedback đã lưu theo đơn — nút hiện "Xem đánh giá" đúng
            com.veganbeauty.app.data.local.OrderReviewLocalStore.applyTo(context, order);

            binding.getRoot().setOnClickListener(v -> onItemClick.onClick(order));

            binding.tvOrderDateTime.setText(order.getOrderDate() + " • " + order.getOrderTime());
            binding.tvOrderId.setText(order.getId());

            binding.tvOrderStatus.setText(order.getStatus());
            int bgResId;
            int textResId;

            switch (order.getStatus()) {
                case "Chờ xác nhận":
                    bgResId = R.color.status_pending_bg;
                    textResId = R.color.status_pending_text;
                    break;
                case "Đang xử lý":
                    bgResId = R.color.status_processing_bg;
                    textResId = R.color.status_processing_text;
                    break;
                case "Đang giao":
                    bgResId = R.color.status_delivering_bg;
                    textResId = R.color.status_delivering_text;
                    break;
                case "Hoàn tất":
                    bgResId = R.color.status_success_bg;
                    textResId = R.color.status_success_text;
                    break;
                case "Đã hủy":
                    bgResId = R.color.status_cancelled_bg;
                    textResId = R.color.status_cancelled_text;
                    break;
                default:
                    bgResId = R.color.status_pending_bg;
                    textResId = R.color.status_pending_text;
                    break;
            }

            binding.tvOrderStatus.getBackground().mutate().setTint(ContextCompat.getColor(context, bgResId));
            binding.tvOrderStatus.setTextColor(ContextCompat.getColor(context, textResId));

            binding.tvProductCountLabel.setText("Sản phẩm (" + order.getItems().size() + ")");

            isExpanded = false;

            updateProductsVisibility(order, context);

            binding.llToggleProducts.setOnClickListener(v -> {
                isExpanded = !isExpanded;
                updateProductsVisibility(order, context);
            });

            String formattedAmount = String.format("%,dđ", order.getTotalAmount()).replace(',', '.');
            binding.tvTotalAmount.setText(formattedAmount);

            setupActionButtons(
                    order,
                    onCancelClick,
                    onDetailClick,
                    onReorderClick,
                    onTrackClick,
                    onContactClick,
                    onReviewClick
            );
        }

        private void updateProductsVisibility(OrderEntity order, Context context) {
            binding.llProductsContainer.removeAllViews();
            
            List<OrderItem> itemsToShow;
            if (isExpanded) {
                itemsToShow = order.getItems();
            } else {
                itemsToShow = order.getItems().size() > 2 ? order.getItems().subList(0, 2) : order.getItems();
            }

            for (OrderItem item : itemsToShow) {
                View rowView = LayoutInflater.from(context).inflate(R.layout.item_account_order_product_row, binding.llProductsContainer, false);

                TextView tvName = rowView.findViewById(R.id.tvProductName);
                TextView tvQuantity = rowView.findViewById(R.id.tvProductQuantity);
                TextView tvPrice = rowView.findViewById(R.id.tvProductPrice);
                ImageView ivImage = rowView.findViewById(R.id.ivProductImage);

                tvName.setText(item.getProductName());
                tvQuantity.setText("x" + item.getQuantity());

                String itemPriceFormatted = String.format("%,dđ", item.getPrice()).replace(',', '.');
                tvPrice.setText(itemPriceFormatted);

                com.bumptech.glide.Glide.with(ivImage.getContext()).load(item.getProductImage()).placeholder(android.R.color.darker_gray).error(android.R.color.darker_gray).into(ivImage);

                binding.llProductsContainer.addView(rowView);
            }

            if (order.getItems().size() > 2) {
                binding.llToggleProducts.setVisibility(View.VISIBLE);
                if (isExpanded) {
                    binding.tvToggleText.setText("Thu gọn");
                    binding.ivToggleIcon.setImageResource(R.drawable.ic_arrow_up);
                } else {
                    binding.tvToggleText.setText("Xem thêm");
                    binding.ivToggleIcon.setImageResource(R.drawable.ic_arrow_down);
                }
            } else {
                binding.llToggleProducts.setVisibility(View.GONE);
            }
        }

        private void setupActionButtons(
                OrderEntity order,
                OnOrderClickListener onCancelClick,
                OnOrderClickListener onDetailClick,
                OnOrderClickListener onReorderClick,
                OnOrderClickListener onTrackClick,
                OnOrderClickListener onContactClick,
                OnOrderClickListener onReviewClick
        ) {
            android.content.Context context = binding.getRoot().getContext();
            binding.btnActionOutlined.setTextColor(ContextCompat.getColor(context, R.color.black));
            binding.btnActionOutlined.setStrokeColorResource(R.color.black);

            switch (order.getStatus()) {
                case "Chờ xác nhận":
                    binding.btnActionOutlined.setVisibility(View.VISIBLE);
                    binding.btnActionOutlined.setText("Hủy đơn");
                    binding.btnActionOutlined.setTextColor(ContextCompat.getColor(context, android.R.color.holo_red_dark));
                    binding.btnActionOutlined.setStrokeColorResource(android.R.color.holo_red_dark);
                    binding.btnActionOutlined.setOnClickListener(v -> onCancelClick.onClick(order));

                    binding.btnActionFilled.setVisibility(View.VISIBLE);
                    binding.btnActionFilled.setText("Liên hệ hỗ trợ");
                    binding.btnActionFilled.setOnClickListener(v -> onContactClick.onClick(order));
                    break;
                case "Đang xử lý":
                    binding.btnActionOutlined.setVisibility(View.VISIBLE);
                    binding.btnActionOutlined.setText("Liên hệ");
                    binding.btnActionOutlined.setOnClickListener(v -> onContactClick.onClick(order));
                    binding.btnActionFilled.setVisibility(View.GONE);
                    break;
                case "Đang giao":
                    binding.btnActionOutlined.setVisibility(View.VISIBLE);
                    binding.btnActionOutlined.setText("Liên hệ");
                    binding.btnActionOutlined.setOnClickListener(v -> onContactClick.onClick(order));
                    binding.btnActionFilled.setVisibility(View.VISIBLE);
                    binding.btnActionFilled.setText("Theo dõi");
                    binding.btnActionFilled.setOnClickListener(v -> onTrackClick.onClick(order));
                    break;
                case "Hoàn tất":
                    binding.btnActionOutlined.setVisibility(View.VISIBLE);
                    binding.btnActionOutlined.setText(order.isHasReview() ? "Xem đánh giá" : "Đánh giá");
                    binding.btnActionOutlined.setOnClickListener(v -> onReviewClick.onClick(order));
                    binding.btnActionFilled.setVisibility(View.VISIBLE);
                    binding.btnActionFilled.setText("Mua lại");
                    binding.btnActionFilled.setOnClickListener(v -> onReorderClick.onClick(order));
                    break;
                case "Đã hủy":
                    binding.btnActionOutlined.setVisibility(View.GONE);
                    binding.btnActionFilled.setVisibility(View.VISIBLE);
                    binding.btnActionFilled.setText("Mua lại");
                    binding.btnActionFilled.setOnClickListener(v -> onReorderClick.onClick(order));
                    break;
                default:
                    binding.btnActionOutlined.setVisibility(View.GONE);
                    binding.btnActionFilled.setVisibility(View.GONE);
                    break;
            }
        }
    }

    public static class OrderDiffCallback extends DiffUtil.ItemCallback<OrderEntity> {
        @Override
        public boolean areItemsTheSame(@NonNull OrderEntity oldItem, @NonNull OrderEntity newItem) {
            return oldItem.getId() != null && oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull OrderEntity oldItem, @NonNull OrderEntity newItem) {
            return oldItem.equals(newItem)
                    && oldItem.isHasReview() == newItem.isHasReview()
                    && oldItem.getReviewStars() == newItem.getReviewStars()
                    && java.util.Objects.equals(oldItem.getReviewText(), newItem.getReviewText());
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/account/order/OrderListAdapter.kt
package com.veganbeauty.app.features.account.order

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import coil.load
import com.veganbeauty.app.R
import com.veganbeauty.app.data.local.entities.OrderEntity
import com.veganbeauty.app.databinding.AccountOrderItemBinding

class OrderListAdapter(
    private val onItemClick: (OrderEntity) -> Unit = {},
    private val onCancelClick: (OrderEntity) -> Unit = {},
    private val onDetailClick: (OrderEntity) -> Unit = {},
    private val onReorderClick: (OrderEntity) -> Unit = {},
    private val onTrackClick: (OrderEntity) -> Unit = {},
    private val onContactClick: (OrderEntity) -> Unit = {},
    private val onReviewClick: (OrderEntity) -> Unit = {}
) : ListAdapter<OrderEntity, OrderListAdapter.OrderViewHolder>(OrderDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): OrderViewHolder {
        val binding = AccountOrderItemBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return OrderViewHolder(binding)
    }

    override fun onBindViewHolder(holder: OrderViewHolder, position: Int) {
        holder.bind(
            getItem(position),
            onItemClick,
            onCancelClick,
            onDetailClick,
            onReorderClick,
            onTrackClick,
            onContactClick,
            onReviewClick
        )
    }

    class OrderViewHolder(private val binding: AccountOrderItemBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(
            order: OrderEntity,
            onItemClick: (OrderEntity) -> Unit,
            onCancelClick: (OrderEntity) -> Unit,
            onDetailClick: (OrderEntity) -> Unit,
            onReorderClick: (OrderEntity) -> Unit,
            onTrackClick: (OrderEntity) -> Unit,
            onContactClick: (OrderEntity) -> Unit,
            onReviewClick: (OrderEntity) -> Unit
        ) {
            val context = binding.root.context

            // Bind root click listener for order detail navigation (phần này làm sau)
            binding.root.setOnClickListener { onItemClick(order) }

            // Bind Date and Time
            binding.tvOrderDateTime.text = "${order.orderDate} • ${order.orderTime}"
            binding.tvOrderId.text = order.id

            // Bind Status Badge
            binding.tvOrderStatus.text = order.status
            val (bgResId, textResId) = when (order.status) {
                "Chờ xác nhận" -> Pair(R.color.status_pending_bg, R.color.status_pending_text)
                "Đang xử lý" -> Pair(R.color.status_processing_bg, R.color.status_processing_text)
                "Đang giao" -> Pair(R.color.status_delivering_bg, R.color.status_delivering_text)
                "Hoàn tất" -> Pair(R.color.status_success_bg, R.color.status_success_text)
                "Đã hủy" -> Pair(R.color.status_cancelled_bg, R.color.status_cancelled_text)
                else -> Pair(R.color.status_pending_bg, R.color.status_pending_text)
            }
            binding.tvOrderStatus.background.mutate().setTint(ContextCompat.getColor(context, bgResId))
            binding.tvOrderStatus.setTextColor(ContextCompat.getColor(context, textResId))

            // Product Container Header
            binding.tvProductCountLabel.text = "Sản phẩm (${order.items.size})"
            
            // State for expand/collapse (defaults to false meaning collapsed)
            var isExpanded = false
            
            fun updateProductsVisibility() {
                binding.llProductsContainer.removeAllViews()
                val itemsToShow = if (isExpanded) order.items else order.items.take(2)
                
                for (item in itemsToShow) {
                    val rowView = LayoutInflater.from(context).inflate(R.layout.item_account_order_product_row, binding.llProductsContainer, false)
                    
                    val tvName = rowView.findViewById<android.widget.TextView>(R.id.tvProductName)
                    val tvQuantity = rowView.findViewById<android.widget.TextView>(R.id.tvProductQuantity)
                    val tvPrice = rowView.findViewById<android.widget.TextView>(R.id.tvProductPrice)
                    val ivImage = rowView.findViewById<android.widget.ImageView>(R.id.ivProductImage)
                    
                    tvName.text = item.productName
                    tvQuantity.text = "x${item.quantity}"
                    
                    val itemPriceFormatted = String.format("%,dđ", item.price).replace(',', '.')
                    tvPrice.text = itemPriceFormatted
                    
                    ivImage.load(item.productImage) {
                        crossfade(true)
                        placeholder(android.R.color.darker_gray)
                        error(android.R.color.darker_gray)
                    }
                    
                    binding.llProductsContainer.addView(rowView)
                }
                
                if (order.items.size > 2) {
                    binding.llToggleProducts.visibility = View.VISIBLE
                    if (isExpanded) {
                        binding.tvToggleText.text = "Thu gọn"
                        binding.ivToggleIcon.setImageResource(R.drawable.ic_chevron_up_thin)
                    } else {
                        binding.tvToggleText.text = "Xem thêm"
                        binding.ivToggleIcon.setImageResource(R.drawable.ic_chevron_down_thin)
                    }
                } else {
                    binding.llToggleProducts.visibility = View.GONE
                }
            }
            
            binding.llToggleProducts.setOnClickListener {
                isExpanded = !isExpanded
                updateProductsVisibility()
            }
            
            // Initial render
            updateProductsVisibility()

            // Formatting Price as 395.000đ
            val formattedAmount = String.format("%,dđ", order.totalAmount).replace(',', '.')
            binding.tvTotalAmount.text = formattedAmount

            // Dynamic Action Buttons
            setupActionButtons(
                order,
                onCancelClick,
                onDetailClick,
                onReorderClick,
                onTrackClick,
                onContactClick,
                onReviewClick
            )
        }

        private fun setupActionButtons(
            order: OrderEntity,
            onCancelClick: (OrderEntity) -> Unit,
            onDetailClick: (OrderEntity) -> Unit,
            onReorderClick: (OrderEntity) -> Unit,
            onTrackClick: (OrderEntity) -> Unit,
            onContactClick: (OrderEntity) -> Unit,
            onReviewClick: (OrderEntity) -> Unit
        ) {
            when (order.status) {
                "Chờ xác nhận" -> {
                    // Outlined: Hủy đơn, Filled: (Gone)
                    binding.btnActionOutlined.visibility = View.VISIBLE
                    binding.btnActionOutlined.text = "Hủy đơn"
                    binding.btnActionOutlined.setOnClickListener { onCancelClick(order) }

                    binding.btnActionFilled.visibility = View.GONE
                }
                "Đang xử lý" -> {
                    // Outlined: Liên hệ, Filled: (Gone)
                    binding.btnActionOutlined.visibility = View.VISIBLE
                    binding.btnActionOutlined.text = "Liên hệ"
                    binding.btnActionOutlined.setOnClickListener { onContactClick(order) }

                    binding.btnActionFilled.visibility = View.GONE
                }
                "Đang giao" -> {
                    // Outlined: Liên hệ, Filled: Theo dõi
                    binding.btnActionOutlined.visibility = View.VISIBLE
                    binding.btnActionOutlined.text = "Liên hệ"
                    binding.btnActionOutlined.setOnClickListener { onContactClick(order) }

                    binding.btnActionFilled.visibility = View.VISIBLE
                    binding.btnActionFilled.text = "Theo dõi"
                    binding.btnActionFilled.setOnClickListener { onTrackClick(order) }
                }
                "Hoàn tất" -> {
                    // Outlined: Đánh giá/Xem đánh giá, Filled: Mua lại
                    binding.btnActionOutlined.visibility = View.VISIBLE
                    binding.btnActionOutlined.text = if (order.hasReview) "Xem đánh giá" else "Đánh giá"
                    binding.btnActionOutlined.setOnClickListener { onReviewClick(order) }

                    binding.btnActionFilled.visibility = View.VISIBLE
                    binding.btnActionFilled.text = "Mua lại"
                    binding.btnActionFilled.setOnClickListener { onReorderClick(order) }
                }

                "Đã hủy" -> {
                    // Outlined: (Gone) - Chi tiết is removed, Filled: Mua lại
                    binding.btnActionOutlined.visibility = View.GONE

                    binding.btnActionFilled.visibility = View.VISIBLE
                    binding.btnActionFilled.text = "Mua lại"
                    binding.btnActionFilled.setOnClickListener { onReorderClick(order) }
                }
                else -> {
                    binding.btnActionOutlined.visibility = View.GONE
                    binding.btnActionFilled.visibility = View.GONE
                }
            }
        }
    }

    class OrderDiffCallback : DiffUtil.ItemCallback<OrderEntity>() {
        override fun areItemsTheSame(oldItem: OrderEntity, newItem: OrderEntity): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: OrderEntity, newItem: OrderEntity): Boolean {
            return oldItem == newItem
        }
    }
}
